aggregator/leader_flush_mgr.go

```diff
@@ -0,0 +1,341 @@
+				bucketIdx: bucketIdx,
+			}
+			mgr.flushTimes.Pop()
+			mgr.flushTimes.Push(nextFlushMetadata)
+			mgr.flushedSincePersist = true
+		} else {
+			// NB(xichen): don't oversleep if the next flush is about to happen.
+			timeToNextFlush := time.Duration(earliestFlush.timeNanos - nowNanos)
+			if timeToNextFlush < waitFor {
+				waitFor = timeToNextFlush
+			}
+		}
+	}
+
+	var (
+		now           = mgr.nowFn()
+		shouldPersist = false
+		flushTimes    schema.ShardSetFlushTimes
+	)
+	durationSinceLastPersist := now.Sub(mgr.lastPersistAt)
+	if mgr.flushedSincePersist && durationSinceLastPersist >= mgr.flushTimesPersistEvery {
+		shouldPersist = true
+		mgr.lastPersistAt = now
+		mgr.flushedSincePersist = false
+		flushTimes = mgr.prepareFlushTimesWithLock(buckets)
+	}
+	mgr.Unlock()
+
+	if !shouldFlush && !shouldPersist {
+		return nil, waitFor
+	}
+	mgr.flushTask.shouldFlush = shouldFlush
+	mgr.flushTask.duration = duration
+	mgr.flushTask.flushers = flushers
+	mgr.flushTask.shouldPersist = shouldPersist
+	mgr.flushTask.flushTimes = &flushTimes
+	return mgr.flushTask, 0
+}
+
+// NB(xichen): if the current instance is a leader, we need to update the flush
+// times heap for the flush goroutine to pick it up.
+func (mgr *leaderFlushManager) OnBucketAdded(
+	bucketIdx int,
+	bucket *flushBucket,
+) {
+	mgr.Lock()
+	mgr.enqueueBucketWithLock(bucketIdx, bucket)
+	mgr.Unlock()
+}
+
+// NB(xichen): leader flush manager can always lead.
+func (mgr *leaderFlushManager) CanLead() bool { return true }
+
+func (mgr *leaderFlushManager) enqueueBucketWithLock(
+	bucketIdx int,
+	bucket *flushBucket,
+) {
+	nowNanos := mgr.nowNanos()
+	flushInterval := bucket.interval
+	nextFlushNanos := nowNanos + int64(flushInterval)
+	if mgr.jitterEnabled {
+		jitterNanos := int64(mgr.rand.Int63n(int64(flushInterval)))
+		nextFlushNanos = nowNanos + jitterNanos
+	}
+	newFlushMetadata := flushMetadata{
+		timeNanos: nextFlushNanos,
+		bucketIdx: bucketIdx,
+	}
+	mgr.flushTimes.Push(newFlushMetadata)
+}
+
+func (mgr *leaderFlushManager) prepareFlushTimesWithLock(
+	buckets []*flushBucket,
+) schema.ShardSetFlushTimes {
+	proto := schema.ShardSetFlushTimes{
+		ByShard: make(map[uint32]*schema.ShardFlushTimes, defaultInitialFlushTimesCapacity),
+	}
+	for _, bucket := range buckets {
+		for _, flusher := range bucket.flushers {
+			shard := flusher.Shard()
+			shardFlushTimes, exists := proto.ByShard[shard]
+			if !exists {
+				shardFlushTimes = &schema.ShardFlushTimes{
+					ByResolution: make(map[int64]int64, 2),
+				}
+				proto.ByShard[shard] = shardFlushTimes
+			}
+			resolution := flusher.Resolution()
+			lastFlushedNanos := flusher.LastFlushedNanos()
+			shardFlushTimes.ByResolution[int64(resolution)] = lastFlushedNanos
+		}
+	}
+	return proto
+}
+
+func (mgr *leaderFlushManager) persistFlushTimes(flushTimes *schema.ShardSetFlushTimes) {
+	persistStart := mgr.nowFn()
+	_, err := mgr.flushTimesStore.Set(mgr.flushTimesKey, flushTimes)
```

Comment:
Should probably retry on error yeah?  Otherwise won't follower need to flush these times?